csrc/multidevice/utils.cpp

```diff
@@ -142,19 +142,49 @@ int64_t numDeviceDims(TensorView* tv) {
 }
 
 bool isResharding(Expr* expr) {
-  std::unordered_set<TensorView*> tvs;
-  for (auto tv : ir_utils::filterByType<TensorView>(expr->inputs())) {
-    tvs.insert(tv);
-  }
-  for (auto tv : ir_utils::filterByType<TensorView>(expr->outputs())) {
-    tvs.insert(tv);
-  }
-  if (tvs.empty()) {
-    return false;
+  // we don't use getTvsWithDifferentSharding because it creates a computeAtMap,
+  // which is too costly
+  for (auto producer : ir_utils::filterByType<TensorView>(expr->inputs())) {
+    for (auto consumer : ir_utils::filterByType<TensorView>(expr->outputs())) {
+      // exit early in the unsharded case for performance
+      if (!producer->hasDeviceMesh() && !consumer->hasDeviceMesh()) {
+        continue;
+      }
+      // If device mesh are different, the Expr is resharding
+      if (producer->getDeviceMesh().vector() !=
+          consumer->getDeviceMesh().vector()) {
+        return true;
+      }
+
+      // Create a map between producer's and consumer's IterDomains. We iterate
+      // over producer's iterdomain and compare sharding type with consumer's
+      // iterdomain
+      const auto p2c_map =
+          PairwiseRootDomainMap(producer, consumer).mapProducerToConsumer();
```

Comment:
This is not equivalent to what's in getTvsWithDifferentSharding. The latter looks at the leaf domain instead of the root domain. While I understand it's probably OK at this moment (given we've been splitting rfactor for DID), I think we'll soon have to fix the discrepancy? Wdyt, @cowanmeg

Reply:
Correct, however I couldn't find any `PairwiseLeafDomainMap`

Reply:
> however I couldn't find any PairwiseLeafDomainMap

Correct -- there's no PariwiseLeafDomainMap. Unfortunately, we probably have to implement it in C++ for loops and branches. I believe this will be needed when we want to split leaf instead for DID.

Reply:
We have a similar problem in `getShardingChanges` since we analyze the root domain for the convenience of PairwiseRootDomainMap.
 
Since this change will help lightning perf, I am good with analyzing the root domain for now and we will fix it later in the leaf+DID PRs!

Reply:
A couple of suggestions:

- Since the problem is `isResharding` is called multiple times at the time of segmentation, it should be possible to do the analysis one and save it for reuse. 
- `ComputeAtMap` and `IdModel` are overkill for this case (although I think we already build one, which may be reusable here). `PairwiseRootDomainMap` and `BestEffortReplay` would be a cheaper alternative. 
- While likely not necessary for multi-device cases yet, this is not a trivial dependency analysis in general, especially with broadcast domains. See, for example, `useSameIndex` in sync_information.cpp. IdModel should greatly simplify the analysis.

Reply:
> Since this change will help lightning perf, I am good with analyzing the root domain for now and we will fix it later in the leaf+DID PRs!

SGTM! 

For the later fix, I'm OK with either changing isResharding to map leaf domains or caching ComputeAtMap/IdModel. I slightly prefer the former because it's stateless. However, if the implementation gets tricky and duplicates what's already in ComputeAtMap/IdModel, I'd try to cache the analysis result as @naoyam suggested, which seems feasible AFAICS from the users of isResharding.